### PR TITLE
ANS-104: Use of NetworkConfig object to get difference between 2 configs.

### PR DESCRIPTION
**Pull request by NehaKembalkarA10**
- Modified `get_diff` method in cliconf plugin by using **NetworkConfig** object in it for getting difference between given 2 configs.
- Added `match` and `running_config` argument's support in `acos_config` module.
- Modified some unit tests to check updated get_diff calls.
- Added unit tests for match argument support.

**Closes:**
ANS-104

### CODE REVIEW

File: examples/acos_config.yml [hunk 1]
```diff
@@ -7,23 +7,21 @@
...
+    - name: configurable backup path without filename
       a10.acos_cli.acos_config:
         backup: yes
-        defaults: true
-        backup_options:
-          filename: acos_default1.cfg
-          dir_path: ./backup_files
 
-    - name: configurable backup path without filename
+    - name: configurable default backup path
```
Comment (hthompson6) [inline examples/acos_config.yml]: name: Backup running config to provided filename

File: examples/acos_config.yml [hunk 2]
```diff
@@ -7,23 +7,21 @@
...
-          - slb template http template1
-          - slb server servertest 20.20.5.147
+          - slb template http template12
+          - slb server servertest 20.20.15.147
           - port 74 tcp
-          - slb server serv1 20.5.7.126
+          - slb server serv13 20.15.27.126
 
-    - name: configurable default backup path
+    - name: configurable backup path without filename
```
Comment (hthompson6) [inline examples/acos_config.yml]: name: Backup running config

File: plugins/action/acos_config.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+__metaclass__ = type
+
+
+from ansible_collections.ansible.netcommon.plugins.action.network import (
+    ActionModule as ActionNetworkModule,
+)
+
+
+class ActionModule(ActionNetworkModule):
+    def run(self, tmp=None, task_vars=None):
```
Comment (hthompson6) [inline plugins/action/acos_config.py]: Needs a space here

File: plugins/cliconf/acos.py [hunk 4]
```diff
@@ -85,6 +86,36 @@ def edit_config(self, candidate=None, commit=True,
...
+        candidate_obj.load(candidate)
+
+        if running and diff_match != 'none':
+            running_obj = NetworkConfig(
+                indent=1, contents=running, ignore_lines=diff_ignore_lines)
+            configdiffobjs = candidate_obj.difference(
+                running_obj, match=diff_match)
+
+        else:
+            configdiffobjs = candidate_obj.items
```
Comment (hthompson6) [inline plugins/cliconf/acos.py]: Change name to `config_diff_objs`

File: plugins/modules/acos_config.py [hunk 5]
```diff
@@ -18,6 +17,8 @@
   - A10 ACOS configurations use a simple block indent file syntax
     for segmenting configuration into sections.  This module provides
     an implementation for working with ACOS configuration sections.
+version_added: '2.9'
+author: Hunter Thompson (@hthompson6)
```
Comment (hthompson6) [inline plugins/modules/acos_config.py]: Let's strip out the author name. Was originally meant as contact point for merging into core. But we are using collections now so it's not as important.

File: plugins/action/acos_config.py [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+        del tmp  # tmp no longer has any effect
+        module_name = self._task.action.split(".")[-1]
+        self._config_module = True if module_name == "acos_config" else False
+        warnings = []
+
+        result = super(ActionModule, self).run(task_vars=task_vars)
+        if warnings:
```
Comment (hthompson6) [inline plugins/action/acos_config.py]: Where is `warnings` being updated? It's not a class attribute and it isn't taking in any function output.

File: tests/unit/modules/network/a10/test_acos_config.py [hunk 7]
```diff
@@ -71,30 +73,68 @@ def test_acos_config_lines(self):
         lines = ["ip dns primary 10.18.18.81"]
         set_module_args(dict(lines=lines))
         self.execute_module()
-        self.conn.edit_config.assert_called_with(
-            candidate=['ip dns primary 10.18.18.81'])
+        self.conn.get_diff = MagicMock(
```
Comment (hthompson6) [inline tests/unit/modules/network/a10/test_acos_config.py]: Why not just use the regular `Mock`? I don't see any magic methods being called upon.

File: tests/unit/modules/network/a10/test_acos_config.py [hunk 8]
```diff
@@ -71,30 +73,68 @@ def test_acos_config_lines(self):
...
+        self.assertTrue(self.conn.edit_config.called)
 
     def test_acos_config_multi_lines(self):
         lines = ["ip dns primary 10.18.18.81", "member rs1-test 80",
                  "slb server server2-test 5.5.5.11"]
         set_module_args(dict(lines=lines))
         self.execute_module()
-        self.conn.edit_config.assert_called_with(candidate=[
-            "ip dns primary 10.18.18.81", "member rs1-test 80", "slb server server2-test 5.5.5.11"])
+        self.conn.get_diff = MagicMock(
```
Comment (hthompson6) [inline tests/unit/modules/network/a10/test_acos_config.py]: (Duplicate comment to make note of similar issue)
Why not just use the regular Mock? I don't see any magic methods being called upon.

File: tests/unit/modules/network/a10/test_acos_config.py [hunk 9]
```diff
@@ -71,30 +73,68 @@ def test_acos_config_lines(self):
...
+        self.assertTrue(self.conn.edit_config.called)
 
     def test_acos_config_before(self):
         lines = ["ip dns primary 10.18.18.19"]
         set_module_args(dict(lines=lines, before=["show avcs"]))
         self.execute_module()
-        self.conn.edit_config.assert_called_with(
-            candidate=['show avcs', 'ip dns primary 10.18.18.19'])
+        commands = ["show avcs", "ip dns primary 10.18.18.19"]
+        self.conn.get_diff = MagicMock(
```
Comment (hthompson6) [inline tests/unit/modules/network/a10/test_acos_config.py]: (Duplicate comment to make note of similar issue)
Why not just use the regular Mock? I don't see any magic methods being called upon.

File: tests/unit/modules/network/a10/test_acos_config.py [hunk 10]
```diff
@@ -71,30 +73,68 @@ def test_acos_config_lines(self):
...
+        self.assertTrue(self.conn.edit_config.called)
 
     def test_acos_config_after(self):
         lines = ["ip dns primary 10.18.18.19"]
         set_module_args(dict(lines=lines, after=["show avcs"]))
         self.execute_module()
-        self.conn.edit_config.assert_called_with(
-            candidate=['ip dns primary 10.18.18.19', 'show avcs'])
+        commands = ["ip dns primary 10.18.18.19", "show avcs"]
+        self.conn.get_diff = MagicMock(
```
Comment (hthompson6) [inline tests/unit/modules/network/a10/test_acos_config.py]: (Duplicate comment to make note of similar issue)
Why not just use the regular Mock? I don't see any magic methods being called upon.

File: plugins/modules/acos_config.py [hunk 11]
```diff
@@ -342,39 +363,33 @@ def main():
...
-
     if module.params['backup'] or (module._diff and
                                    module.params['diff_against'] == 'running'):
         contents = get_config(module, flags=flags)
         if module.params['backup']:
             result['__backup__'] = contents
-            backup(module, contents)
 
-    if module.params['lines']:
+    if any((module.params['lines'], module.params['src'])):
```
Comment (hthompson6) [inline plugins/modules/acos_config.py]: Isn't a need to do this iteration here.

```python
if len(module.params['lines']) > 0 and len(module.params['src']) > 0:
```

`len()` does not iterate over the list. It simply looks up the information stored in memory.

**Review by hthompson6 — CHANGES_REQUESTED** [review]

File: plugins/modules/acos_config.py (hunk 5 above)
Comment (NehaKembalkarA10) [reply plugins/modules/acos_config.py]: Removed author name from all 3 modules

File: plugins/action/acos_config.py (hunk 6 above)
Comment (NehaKembalkarA10) [reply plugins/action/acos_config.py]: There is no need of `warning`. Removed that part of code.

File: plugins/modules/acos_config.py (hunk 11 above)
Comment (NehaKembalkarA10) [reply plugins/modules/acos_config.py]: Yes. Updated that if statement with `if len(str(module.params['lines'])) > 0 and len(str(module.params['src'])) > 0`

File: examples/acos_config.yml (hunk 1 above)
Comment (NehaKembalkarA10) [reply examples/acos_config.yml]: Resolved in PR#7

File: examples/acos_config.yml (hunk 2 above)
Comment (NehaKembalkarA10) [reply examples/acos_config.yml]: Resolved in PR#7

File: plugins/action/acos_config.py (hunk 3 above)
Comment (NehaKembalkarA10) [reply plugins/action/acos_config.py]: Resolved in PR#7

File: plugins/modules/acos_config.py (hunk 5 above)
Comment (NehaKembalkarA10) [reply plugins/modules/acos_config.py]: Resolved in PR#7

File: plugins/action/acos_config.py (hunk 6 above)
Comment (NehaKembalkarA10) [reply plugins/action/acos_config.py]: Resolved in PR#7

File: plugins/modules/acos_config.py (hunk 11 above)
Comment (NehaKembalkarA10) [reply plugins/modules/acos_config.py]: Resolved in PR#7

File: plugins/cliconf/acos.py (hunk 4 above)
Comment (NehaKembalkarA10) [reply plugins/cliconf/acos.py]: Resolved in PR#8

File: tests/unit/modules/network/a10/test_acos_config.py (hunk 7 above)
Comment (NehaKembalkarA10) [reply tests/unit/modules/network/a10/test_acos_config.py]: Yes, here Mock can be used. Replaced `MagicMock` with `Mock` now.
Resolved in PR#8

File: tests/unit/modules/network/a10/test_acos_config.py (hunk 8 above)
Comment (NehaKembalkarA10) [reply tests/unit/modules/network/a10/test_acos_config.py]: Yes, here Mock can be used. Replaced `MagicMock` with `Mock` now.
Resolved in PR#8

File: tests/unit/modules/network/a10/test_acos_config.py (hunk 9 above)
Comment (NehaKembalkarA10) [reply tests/unit/modules/network/a10/test_acos_config.py]: Yes, here Mock can be used. Replaced MagicMock with Mock now.
Resolved in PR#8

File: tests/unit/modules/network/a10/test_acos_config.py (hunk 10 above)
Comment (NehaKembalkarA10) [reply tests/unit/modules/network/a10/test_acos_config.py]: Yes, here Mock can be used. Replaced MagicMock with Mock now.
Resolved in PR#8

**Review by hthompson6 — APPROVED** [review]

**Pull request merged**